packages/vite/src/node/external.ts

```diff
@@ -115,6 +115,16 @@ export function createIsConfiguredAsExternal(
     ) {
       return true
     }
+
+    if (
+      !environment.config.dev.optimizeDeps.noDiscovery &&
+      !environment.config.dev.optimizeDeps.exclude?.includes(id)
+    ) {
+      // If there is server side pre-bundling and the module is not
+      // in the `exclude` config then it is not external
+      return false
+    }
```

Comment:
This one is interesting. We were trying to avoid conflating optimizeDeps.include/exclude with external/noExternal in the past but we will need this check for `noDiscovery` to make sense. This condition should be move to line 140 though.

Reply:
I see, I understand trying not to conflate them, although it does feel like these concepts start to overlap here 🤔 
I hope this is not a huge concern but yeah it does feel like this check is needed since we do not want prebundled modules to be run through `runExternalModule`

In any case I've moved the check 🙂👍

Reply:
> pkg-a will be prebundled with pkg-b, and pkg-a will use the bundled pkg-b. But pkg-c will use the non-bundled pkg-b because pkg-c uses a normal import and Vite cannot intercept that. This means there will be a problem similar to dual package hazard.

Yes I understand this issue, although I imagine that you can always end up in similar situations even today (without environments) if you have multiple packages with the same dependency and exclude one of them, no?
I think this boils down to trusting users to provide correct/sensible configs, if someone does try to use more advanced features such as `optimizeDeps` they should be able to understand the potential risks and implications I think? 🤔

Anyways I don't fully understand your suggestion, dropping this if block means that all the external modules will not be pre-bundled, doesn't it? but that's the whole point of this PR; Allowing environments to get opt-in externals pre-bundling without having to specify all the various dependencies to pre-bundle (since that's something we can do already today).
So I'm not sure what your suggestion is, unless I am missing something?

Reply:
> Yes I understand this issue, although I imagine that you can always end up in similar situations even today (without environments) if you have multiple packages with the same dependency and exclude one of them, no?

Oh, yes, it will happen. But the difference is that we can fix it in future for non-externalized deps since Vite can intercept them and that's not possible for externalized deps. If we add support for prebundling externalized deps, we cannot fix that issue for them. (That said, I'm not so much against keeping this block because I don't think it will be necessarily fixed.)

> dropping this if block means that all the external modules will not be pre-bundled, doesn't it?

Yes.

> but that's the whole point of this PR; Allowing environments to get opt-in externals pre-bundling without having to specify all the various dependencies to pre-bundle (since that's something we can do already today).

I thought the point was to make `noDiscovery: false` work in SSR and allow users / environments to enable it. That would make no-externalized CJS deps work in dev (It should be working in build already). Given that all files has to be bundled for a output for workerd, I thought the ideal way is to set `noExternal: true` (also set `external: ['buffer' /* list supported node and cloudflare builtins */]` if supported) so that all files will be processed by Vite in both dev and build. Do you have any other reason to avoid `ssr.noExternal: true` other than this CJS dep problem?

Reply:
@sapphi-red sorry I simply didn't consider `noExternal: true` + `external: [...]`, it does seems like a valid solution to me, hoping that the maintenance of the `external` list doesn't become too cumbersome 🤔

Yes I agree that this is probably the most correct solution 👍

But, do you think that this could be confusing/easy to get wrong from a user's perspective?
That just setting `noDiscovery: false` is not enough but you also need to set `noExternal: true` for it to take effect

Reply:
(PS: I've removed the `if` block 👍)

Reply:
> hoping that the maintenance of the `external` list doesn't become too cumbersome 🤔

Looking at it again, it seems to work without setting `external`.
https://github.com/vitejs/vite/blob/802839d48335a69eb15f71f2cd816d0b6e4d3556/packages/vite/src/node/plugins/resolve.ts#L459-L480
Maybe It just works only with `noExternal: true`.

> But, do you think that this could be confusing/easy to get wrong from a user's perspective?
> That just setting `noDiscovery: false` is not enough but you also need to set `noExternal: true` for it to take effect

If the runtime does not support running that module natively, then the module should be no-externalized. In other words, if that module can run natively, then it should work without pre-bundle. So I think the user would add the module to `noExternal`.

In future, I guess we'd want to set `noDiscovery: false` by default.